src/lib/sticky-header/sticky-header.ts

```diff
@@ -0,0 +1,321 @@
+    }
+  }
+
+  ngAfterViewInit(): void {
+    if (!this._isStickyPositionSupported) {
+      this.stickyParent = this.parentRegion != null ?
+        this.parentRegion._elementRef.nativeElement : this.element.parentElement;
+      let values = window.getComputedStyle(this.element, '');
+      this._originalStyles = {
+        position: values.position,
+        top: values.top,
+        right: values.right,
+        left: values.left,
+        bottom: values.bottom,
+        width: values.width,
+        zIndex: values.zIndex} as CSSStyleDeclaration;
+      this.attach();
+      this.defineRestrictionsAndStick();
+    }
+  }
+
+  ngOnDestroy(): void {
+    if (this._onScrollSubscription) {
+      this._onScrollSubscription.unsubscribe();
+    }
+
+    if (this._onResizeSubscription) {
+      this._onResizeSubscription.unsubscribe();
+    }
+
+    if (this._onTouchSubscription) {
+      this._onTouchSubscription.unsubscribe();
+    }
+  }
+
+  /**
+   * Check if current browser supports sticky positioning. If yes, apply
+   * sticky positioning. If not, use the original implementation.
+   */
+  setStrategyAccordingToCompatibility(): void {
+    this._isStickyPositionSupported = isPositionStickySupported();
+    if (this._isStickyPositionSupported == true) {
+      this.element.style.top = '0px';
+      this.element.style.cssText += 'position: -webkit-sticky; position: sticky; ';
+      // TODO add css class with both 'sticky' and '-webkit-sticky' on position when @Directory supports adding CSS class
+    }
+  }
+
+  attach() {
+    this._onScrollSubscription = Observable.fromEvent(this.upperScrollableContainer, 'scroll')
+      .debounceTime(DEBOUNCE_TIME).subscribe(() => this.defineRestrictionsAndStick());
+
+    // Have to add a 'onTouchMove' listener to make sticky header work on mobile phones
+    this._onTouchSubscription = Observable.fromEvent(this.upperScrollableContainer, 'touchmove')
+      .debounceTime(DEBOUNCE_TIME).subscribe(() => this.defineRestrictionsAndStick());
+
+    this._onResizeSubscription = Observable.fromEvent(this.upperScrollableContainer, 'resize')
+      .debounceTime(DEBOUNCE_TIME).subscribe(() => this.onResize());
+  }
+
+  onScroll(): void {
+    this.defineRestrictionsAndStick();
+  }
+
+  onTouchMove(): void {
+    this.defineRestrictionsAndStick();
+  }
+
+  onResize(): void {
+    this.defineRestrictionsAndStick();
+    /**
+     * If there's already a header being stick when the page is
+     * resized. The CSS style of the cdkStickyHeader element may be not fit
+     * the resized window. So we need to unstuck it then re-stick it.
+     * unstuck() can set 'isStuck' to FALSE. Then _stickElement() can work.
+     */
+    if (this.isStuck) {
+      this.unstuckElement();
+      this._stickElement();
+    }
+  }
+
+  /**
+   * define the restrictions of the sticky header(including stickyWidth,
+   * when to start, when to finish)
+   */
+  defineRestrictions(): void {
```

Comment:
Can you evaluate each method to make sure their accessor privacy is right? E.g. see which functions need to be public, private, static, etc